cmd/outline-ss-server/metrics.go

```diff
@@ -49,49 +58,164 @@ type outlineMetrics struct {
 var _ service.TCPMetrics = (*outlineMetrics)(nil)
 var _ service.UDPMetrics = (*outlineMetrics)(nil)
 
+// Converts a [net.Addr] to an [IPKey].
+func toIPKey(addr net.Addr, accessKey string) (*IPKey, error) {
+	hostname, _, _ := net.SplitHostPort(addr.String())
+	ip, err := netip.ParseAddr(hostname)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert IP address: %w", err)
+	}
+	return &IPKey{ip, accessKey}, nil
+}
+
+// Represents the clients that are or have been active recently. They stick
+// around until they are inactive, or get reported to Prometheus, whichever
+// comes last.
+type activeClient struct {
+	info         ipinfo.IPInfo
+	connCount    int // The active connection count.
+	startTime    time.Time
+	connDuration time.Duration // If the client has become inactive, this holds the connection duration.
+}
+
+type IPKey struct {
+	ip        netip.Addr
+	accessKey string
+}
+
+type tunnelTimeCollector struct {
+	ipinfo.IPInfoMap
+	mu            sync.Mutex // Protects the activeClients map.
+	activeClients map[IPKey]*activeClient
+
+	tunnelTimePerKey      *prometheus.Desc
```

Comment:
I don't know if it makes your life easier, but it's possible to implement with two CounterVecs here, then call their Describe methods in tunnelTimeCollector.Describe, and their Collect metrics in tunnelTimeCollector.Collect, after you calculated the changes and made the Add calls.

Having your own maps might be more efficient though.

Reply:
I just tried it and I like it. It's cleaner and makes it clear what's happening. After looking at how prometheus implemented the `MetricVec`s, this is pretty similar. Thanks for the suggestion.